tests/csapi/sync_archive_test.go

```diff
@@ -0,0 +1,436 @@
+
+	roomLeaveRes := syncRes.Get("rooms.leave")
+	if roomLeaveRes.Exists() {
+		if !roomLeaveRes.IsObject() {
+			t.Fatalf("rooms.leave is not an object in syncRes")
+		}
+
+		if len(roomLeaveRes.Map()) != 0 {
+			t.Fatalf("Expected no rooms in 'leave' state")
+		}
+	}
+}
+
+// sytest: We should see our own leave event, even if history_visibility is
+func TestLeaveEventVisibility(t *testing.T) {
+	runtime.SkipIf(t, runtime.Dendrite) // FIXME: https://github.com/matrix-org/dendrite/issues/1323
+
+	// Note: this deviates from the sytest by introducing a second user,
+	//  this user is only meant to keep the room alive,
+	//  as a room with no users may be purged by the server,
+	//  creating side effects that this test is not looking for.
+	deployment := Deploy(t, b.BlueprintOneToOneRoom)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+
+	aliceFilter := createFilter(t, alice, map[string]interface{}{
+		"room": map[string]interface{}{
+			"include_leave": true,
+		},
+	})
+
+	roomID := alice.CreateRoom(t, map[string]interface{}{
+		"initial_state": []map[string]interface{}{
+			{
+				"content": map[string]interface{}{
+					"history_visibility": "joined",
+				},
+				"type":      "m.room.history_visibility",
+				"state_key": "",
+			},
+		},
+		"preset": "public_chat",
+	})
+	bob.JoinRoom(t, roomID, nil)
+
+	aliceSince := alice.MustSyncUntil(
+		t,
+		client.SyncReq{Filter: aliceFilter},
+		client.SyncJoinedTo(alice.UserID, roomID),
+		client.SyncJoinedTo(bob.UserID, roomID),
+	)
+
+	alice.LeaveRoom(t, roomID)
+	bob.MustSyncUntil(t, client.SyncReq{}, client.SyncLeftFrom(alice.UserID, roomID))
+
+	syncRes, _ := alice.MustSync(t, client.SyncReq{Filter: aliceFilter, Since: aliceSince, TimeoutMillis: "0"})
+
+	roomRes := syncRes.Get("rooms.leave." + client.GjsonEscape(roomID))
+
+	if !roomRes.Exists() {
+		t.Fatalf("could not find %s in rooms.leave", roomID)
+	}
+
+	stateResult := roomRes.Get("state.events")
+
+	if stateResult.Exists() {
+		if !stateResult.IsArray() {
+			t.Fatalf("state.events was not an array")
+		} else {
+			if len(stateResult.Array()) != 0 {
+				t.Fatalf("Expected no state events")
+			}
+		}
+	}
+
+	timelineRes := roomRes.Get("timeline.events")
+
+	if !timelineRes.Exists() {
+		t.Fatalf("could not find timeline.events")
+	} else if !timelineRes.IsArray() {
+		t.Fatalf("timeline.events was not an array")
+	} else if len(timelineRes.Array()) != 1 {
+		t.Fatalf("expected exactly 1 item in timeline")
+	}
+
+	event := timelineRes.Array()[0]
+
+	if event.Get("type").Str != "m.room.member" {
+		t.Fatalf("expected timeline event was not m.room.member")
+	} else if event.Get("sender").Str != alice.UserID {
+		t.Fatalf("sender of membership event was not alice")
+	} else if event.Get("content.membership").Str != "leave" {
+		t.Fatalf("membership event was not leave")
+	}
+}
+
+// sytest: We should see our own leave event when rejecting an invite,
```

Comment:
typo:
```suggestion
// sytest: We should see our own leave event when rejecting an invite, even if history_visibility is restricted (riot-web/3462)
```

Reply:
Ditto to my previous comment, not a typo